jubatus/core/classifier/nearest_neighbor_classifier.cpp

```diff
@@ -75,13 +75,21 @@ void nearest_neighbor_classifier::train(
   }
   if (unlearner_) {
     util::concurrent::scoped_lock unlearner_lk(unlearner_mutex_);
+
+    // acquire the lock outside of touch() function
+    shared_ptr<storage::column_table> table =
+        nearest_neighbor_engine_->get_table();
+    util::concurrent::scoped_wlock lk(table->get_mutex());
+
     if (!unlearner_->touch(id)) {
       throw JUBATUS_EXCEPTION(common::exception::runtime_error(
           "cannot add new ID as number of sticky IDs reached "
           "the maximum size of unlearner: " + id));
     }
   }
-  nearest_neighbor_engine_->set_row(id, fv);
+
+  // unlearner is not called in set_row
+  nearest_neighbor_engine_->set_row(id, fv);  // lock acquired inside
```

Comment:
I think ``scoped_wlock lk`` is used only in the if-statement scope of the following:
https://github.com/jubatus/jubatus_core/pull/303/files#diff-a69d8971d46596cf55d633ea1cbeb970R76